src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs

```diff
@@ -224,6 +228,57 @@ private void RetrieveNavigationSourcePaths(IEdmNavigationSource navigationSource
             Debug.Assert(path.Any() == false);
         }
 
+        /// <summary>
+        /// Retrieves the paths for properties of type complex type from entities
+        /// </summary>
+        /// <param name="entityType">The entity type.</param>
+        /// <param name="currentPath">The current path.</param>
+        /// <param name="convertSettings">The settings for the current conversion.</param>
+        private void RetrieveComplexPropertyPaths(IEdmEntityType entityType, ODataPath currentPath, OpenApiConvertSettings convertSettings)
+        {
+            Debug.Assert(entityType != null);
+            Debug.Assert(currentPath != null);
+            Debug.Assert(convertSettings != null);
+
+            if (!convertSettings.EnableNavigationPropertyPath) return;
+
+            foreach (IEdmStructuralProperty sp in entityType.StructuralProperties()
+                                                    .Where(x => x.Type.IsComplex() ||
+                                                            x.Type.IsCollection() && x.Type.Definition.AsElementType() is IEdmComplexType))
+            {
+                currentPath.Push(new ODataComplexPropertySegment(sp));
+                AppendPath(currentPath.Clone());
+
+
+                if (sp.Type.IsCollection())
+                {
+                    CreateTypeCastPaths(currentPath, convertSettings, sp.Type.Definition.AsElementType() as IEdmComplexType, sp, true);
+                    var count = _model.GetRecord<CountRestrictionsType>(sp, CapabilitiesConstants.CountRestrictions);
+                    if(count?.IsCountable ?? true)
+                        CreateCountPath(currentPath, convertSettings);
+                }
+                else
+                {
+                    var complexTypeReference = sp.Type.AsComplex();
+                    var definition = complexTypeReference.ComplexDefinition();
+
+                    CreateTypeCastPaths(currentPath, convertSettings, definition, sp, false);
+                    foreach (IEdmNavigationProperty np in complexTypeReference
+                                                        .DeclaredNavigationProperties()
+                                                        .Union(definition
+                                                                        .FindAllBaseTypes()
+                                                                        .SelectMany(x => x.DeclaredNavigationProperties()))
+                                                        .Distinct()
+                                                        .Where(CanFilter))
+                    {
+                        var count = _model.GetRecord<CountRestrictionsType>(np, CapabilitiesConstants.CountRestrictions);
+                        RetrieveNavigationPropertyPaths(np, count, currentPath, convertSettings);
```

Comment:
`RetrieveNavigationPropertyPaths` will throw an exception when calling [`ShouldExpandNavigationProperty`](https://github.com/microsoft/OpenAPI.NET.OData/blob/e647a1be01598b0ea9acb41325e8e47a8bf8fcaa/src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs#L339) since[ `OdataSegment.EntityType`](https://github.com/microsoft/OpenAPI.NET.OData/blob/e647a1be01598b0ea9acb41325e8e47a8bf8fcaa/src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs#L443) is not yet implemented:
https://github.com/microsoft/OpenAPI.NET.OData/blob/e647a1be01598b0ea9acb41325e8e47a8bf8fcaa/src/Microsoft.OpenApi.OData.Reader/Edm/ODataSegment.cs#L90

Reply:
Not sure if it's related but, a null exception is thrown when attempting to retrieve the "Entity type of a Complex type property segment" here:
https://github.com/microsoft/OpenAPI.NET.OData/blob/9255901b9d43b08e4a68f330bb8ace9c5bac5bad/src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs#L443

Example:
The complex property segment in question:
![image](https://user-images.githubusercontent.com/40403681/149950035-de7995c4-1610-494f-b074-85c530c2a44a.png)

The complex type which has containment navigation properties
![image](https://user-images.githubusercontent.com/40403681/149949744-46c4e18c-72a0-495c-8453-61c42bc65a35.png)

Reply:
I've just pushed another commit that should solve your issue, please pull and give it another try.